rig-postgres/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+              SELECT DISTINCT ON (id) id, document, embedding {} $1 as distance \
+              FROM {} \
+              ORDER BY id, distance \
+            ) \
+            ORDER BY distance \
+            LIMIT $2",
+            self.distance_function, self.documents_table
+        )
+    }
+
+    pub async fn insert_documents<Doc: Serialize + Embed + Send>(
+        &self,
+        documents: Vec<(Doc, OneOrMany<Embedding>)>,
+    ) -> Result<(), VectorStoreError> {
+        for (document, embeddings) in documents {
+            let id = Uuid::new_v4();
+            let json_document = serde_json::to_value(&document).unwrap();
+
+            for embedding in embeddings {
+                let embedding_text = embedding.document;
+                let embedding: Vec<f64> = embedding.vec;
+
+                sqlx::query(
+                    format!(
+                        "INSERT INTO {} (id, document, embedded_text, embedding) VALUES ($1, $2, $3, $4)",
+                        self.documents_table
+                    )
+                    .as_str(),
+                )
+                .bind(id)
+                .bind(&json_document)
+                .bind(&embedding_text)
+                .bind(&embedding)
+                .execute(&self.pg_pool)
+                .await
+                .map_err(|e| VectorStoreError::DatastoreError(e.into()))?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<Model: EmbeddingModel> VectorStoreIndex for PostgresVectorStore<Model> {
+    /// Get the top n documents based on the distance to the given query.
+    /// The result is a list of tuples of the form (score, id, document)
+    async fn top_n<T: for<'a> Deserialize<'a> + Send>(
+        &self,
+        query: &str,
+        n: usize,
+    ) -> Result<Vec<(f64, String, T)>, VectorStoreError> {
+        let embedded_query: pgvector::Vector = self
+            .model
+            .embed_text(query)
+            .await?
+            .vec
+            .iter()
+            .map(|&x| x as f32)
+            .collect::<Vec<f32>>()
+            .into();
+
+        let rows: Vec<SearchResult> = sqlx::query_as(self.search_query().as_str())
+            .bind(embedded_query)
+            .bind(n as i64)
+            .fetch_all(&self.pg_pool)
+            .await
+            .map_err(|e| VectorStoreError::DatastoreError(Box::new(e)))?;
+
+        let rows: Vec<(f64, String, T)> = rows
+            .into_iter()
+            .flat_map(SearchResult::into_result)
+            .collect();
+
+        Ok(rows)
+    }
+
+    /// Same as `top_n` but returns the document ids only.
+    async fn top_n_ids(
+        &self,
+        query: &str,
+        n: usize,
+    ) -> Result<Vec<(f64, String)>, VectorStoreError> {
+        self.top_n::<Value>(query, n).await.map(|result| {
+            result
+                .into_iter()
+                .map(|(distance, id, _)| (distance, id))
+                .collect()
+        })
+    }
```

Comment:
This still fetches the documents over the wire (which may or may not be a lot), what do you think of using an entirely different query that only selects the scores and ids? Might be premature optimization, so not sure if that's worth the added complexity

Reply:
It has sense and it also remove the deserialization step (not needed).
I added the new logic and updated integration test.